enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,475 @@
+- Storage CSI migration will be required for OpenStack and vSphere providers
+- Networking:
+  - Q: In the CCM interface the `Routes` [interface](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L63) is implemented. Does it have any relation to the SDN? [Link](https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go) to the controller managing this.
+  A: No, the logic for `Routes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default across all providers.
+  - Q: Load Balancer setup for `Service` resources is [supported](https://github.com/kubernetes/cloud-provider/blob/master/controllers/service/controller.go) by the CCM interface with the [LoadBalancer()](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L49] method. How important is this functionality during cluster lifetime? 
+  A: Only the [ingress-operator](https://github.com/openshift/cluster-ingress-operator) is affected by this. Any `Route` resource changes will not be configured while the `CCM` is down, and any request to provision an external access to the cluster will not be satisfied.
+
+3. Should we reuse the existing cloud provider config or generate a new one?
+CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+4. How to migrate PVs created by the in-tree cloud provider?
+[CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
+
+## Summary
+
+This enhancement proposal describes the migration of cloud platforms from the deprecated [in-tree cloud providers](https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#openstack) to [Cloud Controller Manager](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-openstack-cloud-controller-manager.md#get-started-with-external-openstack-cloud-controller-manager-in-kubernetes) services that implement `external cloud provider` [interface](https://github.com/kubernetes/cloud-provider).
+
+As a pioneer platform, it is proposed to use OpenStack.
+
+## Motivation
+
+Using Cloud Controller Managers (CCMs) is the Kubernetes' [preferred way](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/) to interact with underlying cloud platforms as it provides more flexibility and freedom for developers. It replaces existing in-tree cloud providers, which have been deprecated and will be permanently removed approximately in Kubernetes [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). But they are still used in OpenShift and we must start a smooth migration towards CCMs.
```

Comment:
> approximately in Kubernetes [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).

that statement is from May; the current target is 1.23

Reply:
Hi Dan! Thank you for the information!
Can you provide any links to confirm that the removal has been delayed, so we can include them in the proposal?

Reply:
I think it is related to the removal of the field completely in kubelet in 1.23 - https://github.com/kubernetes/kubernetes/pull/90408#issuecomment-670057176 Let me have a better look.

Reply:
sorry, missed this question earlier. I don't have a link to an official source. Andrew Sy Kim (who made the Slack comment you linked to) mentioned the 1.23 target [here](https://github.com/kubernetes/enhancements/pull/1665#issuecomment-703140901) and then I asked him on Slack and he confirmed that's the current target, and it's changed since the conversation you linked to.

Reply:
Thanks! I'm adding the updated info to the proposal